File: examples/rust/src/bin/bit_mask_operator.rs
use tsl_generated::tsl_core::StaticSimdVector;
use tsl_generated::tsl_scalar as tsl;

struct LessThan;

impl<V> tsl::algo::BinaryPredicateKernel<V> for LessThan
where
    V: StaticSimdVector<BaseType = i32> + tsl::detail::primitives::Less_thanImpl,
{
    fn test(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::MaskType {
        tsl::less_than::<V>(left, right)
    }
}

struct SquareWhere;

impl<V> tsl::algo::MaskedUnaryKernel<V> for SquareWhere
where
    V: StaticSimdVector<BaseType = i32> + tsl::detail::primitives::MulImpl,
    V::RegisterType: Copy,
{
    fn apply(&mut self, _active: V::MaskType, value: V::RegisterType) -> V::RegisterType {
        tsl::mul::<V>(value, value)
    }
}

struct AddOrLeft;

impl<V> tsl::algo::MaskedBinaryKernel<V> for AddOrLeft
where
    V: StaticSimdVector<BaseType = i32>
        + tsl::detail::primitives::AddImpl
        + tsl::detail::primitives::BlendImpl,
    V::RegisterType: Copy,
{
    fn apply(
        &mut self,
        active: V::MaskType,
        left: V::RegisterType,
        right: V::RegisterType,
    ) -> V::RegisterType {
        let sum = tsl::add::<V>(left, right);
        tsl::blend::<V>(active, left, sum)
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = ((i * 3) % 37) as i32 - 18;
        *right_value = ((i * 17) % 53) as i32 - 26;
    }
}

fn packed_mask_active(masks: &[u8], row: usize) -> bool {
    ((masks[row / 8] >> (row % 8)) & 1) != 0
}

fn main() {
    const COUNT: usize = 1003;

    let mut left = vec![0i32; COUNT];
    let mut right = vec![0i32; COUNT];
    fill_inputs(&mut left, &mut right);

    macro_rules! run_policy {
        ($policy:expr) => {{
            let policy = $policy;
            let mask_count = tsl::algo::bit_mask_count::<_, i32>(policy, left.len());
            let mut masks = vec![0xFFu8; mask_count];

            let mut predicate = LessThan;
            let produced = tsl::algo::predicate_binary_mask_layout::<
                _,
                tsl::algo::mask_layout::Bits,
                _,
                i32,
            >(policy, &mut predicate, &left, &right, &mut masks);
            assert_eq!(produced, masks.len());

            for i in 0..left.len() {
                assert_eq!(packed_mask_active(&masks, i), left[i] < right[i]);
            }
            for i in left.len()..(masks.len() * 8) {
                assert!(!packed_mask_active(&masks, i));
            }

            let preserved = -567890;
            let mut unary_output = vec![preserved; left.len()];
            let mut square = SquareWhere;
            tsl::algo::transform_where_unary_mask_layout::<
                _,
                tsl::algo::mask_layout::Bits,
                _,
                i32,
            >(policy, &mut square, &left, &masks, &mut unary_output);

            for (i, actual) in unary_output.iter().enumerate() {
                let expected = if left[i] < right[i] {
                    left[i] * left[i]
                } else {
                    preserved
                };
                assert_eq!(*actual, expected);
            }

            let mut binary_output = vec![0i32; left.len()];
            let mut add = AddOrLeft;
            tsl::algo::transform_masked_binary_mask_layout::<
                _,
                tsl::algo::mask_layout::Bits,
                _,
                i32,
            >(policy, &mut add, &left, &right, &masks, &mut binary_output);

            for (i, actual) in binary_output.iter().enumerate() {
                let expected = if left[i] < right[i] {
                    left[i] + right[i]
                } else {
                    left[i]
                };
                assert_eq!(*actual, expected);
            }
        }};
    }

    run_policy!(tsl::algo::parallelism::native());
    run_policy!(tsl::algo::parallelism::fixed::<1>());
    run_policy!(tsl::algo::parallelism::generic::<4>());
    run_policy!(tsl::algo::parallelism::generic::<16>());
}

File: examples/cpp/bit_mask_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct less_than_op {
    template <class Vec>
    typename Vec::mask_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::less_than<Vec>(left, right);
    }
};

struct square_where_op {
    template <class Vec>
    typename Vec::register_type operator()(
        typename Vec::mask_type active,
        typename tsl::reg_param<Vec>::type value) const {
        (void)active;
        return tsl::mul<Vec>(value, value);
    }
};

struct add_or_left_op {
    template <class Vec>
    typename Vec::register_type operator()(
        typename Vec::mask_type active,
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        const auto sum = tsl::add<Vec>(left, right);
        return tsl::select<Vec>(active, sum, left);
    }
};

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>((i * 3) % 37) - 18);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 17) % 53) - 26);
    }
}

bool packed_mask_active(const std::vector<std::uint8_t>& masks, std::size_t row) {
    return ((masks[row / 8] >> (row % 8)) & std::uint8_t{1}) != 0;
}

template <class Parallelism>
bool run_bit_where_case() {
    constexpr std::size_t count = 1003;
    constexpr std::int32_t preserved = -567890;
    std::vector<std::int32_t> input(count);
    std::vector<std::int32_t> threshold(count);
    std::vector<std::int32_t> output(count, preserved);
    fill_inputs(input, threshold);

    using mask_type = tsl::algo::bit_mask_type<Parallelism, std::int32_t>;
    std::vector<mask_type> masks(
        tsl::algo::bit_mask_count<Parallelism, std::int32_t>(count),
        std::uint8_t{0xFF});

    const auto produced = tsl::algo::predicate_binary<
        Parallelism,
        tsl::algo::alignment::unaligned,
        tsl::algo::mask_layout::bits>(
        less_than_op{},
        input.data(),
        threshold.data(),
        masks.data(),
        count);
    if (produced != masks.size()) {
        return false;
    }

    for (std::size_t i = 0; i < count; ++i) {
        const bool expected = input[i] < threshold[i];
        if (packed_mask_active(masks, i) != expected) {
            return false;
        }
    }
    for (std::size_t i = count; i < masks.size() * 8; ++i) {
        if (packed_mask_active(masks, i)) {
            return false;
        }
    }

    tsl::algo::transform_where_unary<
        Parallelism,
        tsl::algo::alignment::unaligned,
        tsl::algo::mask_layout::bits>(
        square_where_op{},
        input.data(),
        masks.data(),
        output.data(),
        count);

    for (std::size_t i = 0; i < count; ++i) {
        const bool active = input[i] < threshold[i];
        const auto expected = active ? input[i] * input[i] : preserved;
        if (output[i] != expected) {
            return false;
        }
    }
    return true;
}

template <class Parallelism>
bool run_bit_masked_case() {
    constexpr std::size_t count = 1003;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    std::vector<std::int32_t> output(count, 0);
    fill_inputs(left, right);

    using mask_type = tsl::algo::bit_mask_type<Parallelism, std::int32_t>;
    std::vector<mask_type> masks(
        tsl::algo::bit_mask_count<Parallelism, std::int32_t>(count));

    tsl::algo::predicate_binary<
        Parallelism,
        tsl::algo::alignment::unaligned,
        tsl::algo::mask_layout::bits>(
        less_than_op{},
        left.data(),
        right.data(),
        masks.data(),
        count);
    tsl::algo::transform_masked_binary<
        Parallelism,
        tsl::algo::alignment::unaligned,
        tsl::algo::mask_layout::bits>(
        add_or_left_op{},
        left.data(),
        right.data(),
        masks.data(),
        output.data(),
        count);

    for (std::size_t i = 0; i < count; ++i) {
        const bool active = left[i] < right[i];
        const auto expected = active ? left[i] + right[i] : left[i];
        if (output[i] != expected) {
            return false;
        }
    }
    return true;
}

template <class Parallelism>
bool run_bit_mask_cases() {
    return run_bit_where_case<Parallelism>() &&
           run_bit_masked_case<Parallelism>();
}

int main() {
    if (!run_bit_mask_cases<tsl::dataparallel::fixed<1>>()) {
        return 1;
    }
    if (!run_bit_mask_cases<tsl::dataparallel::generic<4>>()) {
        return 2;
    }
    if (!run_bit_mask_cases<tsl::dataparallel::generic<16>>()) {
        return 3;
    }
    return 0;
}
